### ISSUE TITLE
Make Auto Volume Growth Configurable

### ISSUE THREAD

**Issue by chrislusf** [seaweedfs/seaweedfs#6210]
### Discussed in https://github.com/seaweedfs/seaweedfs/discussions/6157

<div type='discussions-op-text'>

<sup>Originally posted by **alejandrorascovan** October 22, 2024</sup>
Currently, the settings for automatic volume growth appear to be hard-coded in the project. It would be great to make these settings configurable externally.

For instance, in our use case, we work with relatively small volume sizes (around 1GB). If all volumes fill up in less than 14 minutes, which seems to be the current [interval](https://github.com/seaweedfs/seaweedfs/blob/master/weed/server/master_grpc_server_volume.go#L51) at which the auto volume growth routine runs, there would be no writable volumes available during this time window.

It would be great to have the ability to configure both the auto volume growth interval and how many volumes are added during each growth cycle.</div>

**Comment by sek2022** [comment]
I think free configuration is reasonable, Like this：
https://github.com/sek2022/seaweedfs/commit/0796ee23141b346ac73d787fbf9e7c519b17e819#diff-05fd1cf805ee6362800cc934fd8dd8e1349a6419be692698f6d33b4ec688a1afR74

**Comment by kmlebedev** [comment]
> I think free configuration is reasonable, Like this： [sek2022@0796ee2#diff-05fd1cf805ee6362800cc934fd8dd8e1349a6419be692698f6d33b4ec688a1afR74](https://github.com/sek2022/seaweedfs/commit/0796ee23141b346ac73d787fbf9e7c519b17e819#diff-05fd1cf805ee6362800cc934fd8dd8e1349a6419be692698f6d33b4ec688a1afR74)

Thank you for bringing this issue to my attention. And here I need to think carefully about the balance, and
For now I consider this commit harmful and wrote in the review why.

**Comment by kmlebedev** [comment]
@sek2022 Tell us your specific case, what is your problem, maybe you are solving it with the wrong tool.
For example, minWriteableVolumeSize is the volumeGrowthCount for a specific collection/bucket and you can set it like this.

The interval is also not so important, since without it the main mechanism operates, which creates new volumes with each request.

**Comment by sek2022** [comment]
> @sek2022 Tell us your specific case, what is your problem, maybe you are solving it with the wrong tool. For example, minWriteableVolumeSize is the volumeGrowthCount for a specific collection/bucket and you can set it like this.
> 
> The interval is also not so important, since without it the main mechanism operates, which creates new volumes with each request. And we don't expect Auto Volume Growth to work.
> 
> The main task is to stretch the process of volume growth over time so that the volumes gradually and smoothly flow into the crowded, and not all at once. If this task is not achieved in your case, then let's discuss how to improve it.



> @sek2022 Tell us your specific case, what is your problem, maybe you are solving it with the wrong tool. For example, minWriteableVolumeSize is the volumeGrowthCount for a specific collection/bucket and you can set it like this.
> 
> The interval is also not so important, since without it the main mechanism operates, which creates new volumes with each request. And we don't expect Auto Volume Growth to work.
> 
> The main task is to stretch the process of volume growth over time so that the volumes gradually and smoothly flow into the crowded, and not all at once. If this task is not achieved in your case, then let's discuss how to improve it.

@kmlebedev  My requirement is to store 600G of data per minute, so the writing speed of 6 volumes is far from meeting the requirement. I need at least 80 volumes to write at the same time to meet the requirement, so there are the above improvements.
I have tried to use volume.grow, But It does not solve my problem, Each execution can only temporarily add a few volumes, Volumes will soon be insufficient.
So I need this:https://github.com/sek2022/seaweedfs/commit/0796ee23141b346ac73d787fbf9e7c519b17e819#diff-05fd1cf805ee6362800cc934fd8dd8e1349a6419be692698f6d33b4ec688a1afR74

**Comment by kmlebedev** [comment]
@sek2022 You misunderstood me, I meant the setting
```
weed scaffold -config=master
# create this number of logical volumes if no more writable volumes
# count_x means how many copies of data.
# e.g.:
#   000 has only one copy, copy_1
#   010 and 001 has two copies, copy_2
#   011 has only 3 copies, copy_3
[master.volume_growth]
copy_1 = 7                # create 1 x 7 = 7 actual volumes
copy_2 = 6                # create 2 x 6 = 12 actual volumes
copy_3 = 3                # create 3 x 3 = 9 actual volumes
copy_other = 1            # create n x 1 = n actual volumes
```

What value are you currently using? 
This is similar to your minWriteableVolumeSize.

**Comment by sek2022** [comment]
> @sek2022 You misunderstood me, I meant the setting
> 
> ```
> weed scaffold -config=master
> # create this number of logical volumes if no more writable volumes
> # count_x means how many copies of data.
> # e.g.:
> #   000 has only one copy, copy_1
> #   010 and 001 has two copies, copy_2
> #   011 has only 3 copies, copy_3
> [master.volume_growth]
> copy_1 = 7                # create 1 x 7 = 7 actual volumes
> copy_2 = 6                # create 2 x 6 = 12 actual volumes
> copy_3 = 3                # create 3 x 3 = 9 actual volumes
> copy_other = 1            # create n x 1 = n actual volumes
> ```
> 
> What value are you currently using? This is similar to your minWriteableVolumeSize.

Yes, You are right! I didn't notice this can configurable in master.toml. It's hidden.

**Comment by sek2022** [comment]
I think one value is enough for master.volume_growth, It is related to the write speed and has nothing to do with the copy number.

**Comment by kmlebedev** [comment]
In this PR I reduced it to 5 minutes, this should be enough to support the required number of writeble volumes from those specified master.volume_growth by you in the config.

With this volume you need to use the large version where the houses can be up to 800Gb. 
It seems you have chosen the wrong tools.

https://github.com/seaweedfs/seaweedfs/pull/6214
